# 05 Grouping & Aggregating Variables

    01 Import merged and derived data
    02 Aggregated mean of the “order_number” column grouped by “department_id” 
    03 Create loyalty flag for existing customers 
    04 Descriptive statistics by loyalty
    05 Create spending_flag
    06 Create order_frequency flag
    07 Export data

In [1]:
# import libraries
import pandas as pd
import numpy as np
import os

## 01 Import data

In [2]:
# Creating a path
path = r'C:\Users\Lenovo\OneDrive\Dokumente\CareerFoundry\Data Immersion\4. Python Fundamentals for Data Analysts\Instacart Basket Analysis - Kopie'

In [3]:
# import orders_products_merged_derived
opm=pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_merge_derived.pkl'))

In [4]:
# Creating a subset of the first million rows
df=opm[:1000000]


In [5]:
# Checking shape
df.shape

(1000000, 19)

In [6]:
# Checking head
opm.head(10)

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   
5           1  Chocolate Sandwich Cookies        61             19     5.8   
6           1  Chocolate Sandwich Cookies        61             19     5.8   
7           1  Chocolate Sandwich Cookies        61             19     5.8   
8           1  Chocolate Sandwich Cookies        61             19     5.8   
9           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   3139998      138            28                   6                 11   
1   1977647      138            30                   6                 17   
2    389851      709             2                   0                 21   
3    652770      764             1                   3                 13   
4   1813452      764             3                   4                 17   
5   1701441      777            16                   1                  7   
6   1871483      825             3                   2                 14   
7   1290456      910            12                   3                 10   
8    369558     1052            10                   1                 20   
9    589712     1052            15                   1                 12   

   days_since_prior_order  new_customer  add_to_cart_order  reordered _merge  \
0                     3.0         False                  5          0   both   
1                    20.0         False                  1          1   both   
2                     6.0         False                 20          0   both   
3                     NaN          True                 10          0   both   
4                     9.0         False                 11          1   both   
5                    26.0         False                  7          0   both   
6                    30.0         False                  2          0   both   
7                    30.0         False                  1          0   both   
8                    19.0         False                  1          0   both   
9                    15.0         False                  2          1   both   

     price_range_loc     busiest_day    busiest_days   busiest_hours  
0  Mid-range product  Regularly busy  Regularly busy     Most orders  
1  Mid-range product  Regularly busy  Regularly busy  Average orders  
2  Mid-range product     Busiest day    Busiest days  Average orders  
3  Mid-range product  Regularly busy    Slowest days     Most orders  
4  Mid-range product      Least busy    Slowest days  Average orders  
5  Mid-range product  Regularly busy    Busiest days  Average orders  
6  Mid-range product  Regularly busy  Regularly busy     Most orders  
7  Mid-range product  Regularly busy    Slowest days     Most orders  
8  Mid-range product  Regularly busy    Busiest days  Average orders  
9  Mid-range product  Regularly busy    Busiest days     Most orders

In [7]:
# Group by products
df.groupby('product_name')

In [8]:
# Grouping by products and calculating the average number of orders
df.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.577493
2                17.320781
3                16.084944
4                17.530458
5                14.763075
6                16.658449
7                17.031590
8                15.076662
9                15.447580
10               18.681852
11               15.447411
12               14.327957
13               16.548642
14               16.960241
15               16.121948
16               17.803851
17               15.593633
18               19.674252
19               16.899756
20               16.255442
21               25.535479

In [9]:
# Checking the sahpe
df.shape

(1000000, 19)

In [10]:
# repeat the calculation, adding min and max number of orders
df.groupby('department_id').agg({'order_number': ['mean', 'min', 'max']})

order_number        
                      mean min max
department_id                     
1                15.577493   1  99
2                17.320781   1  96
3                16.084944   1  99
4                17.530458   1  99
5                14.763075   1  99
6                16.658449   1  99
7                17.031590   1  99
8                15.076662   1  98
9                15.447580   1  99
10               18.681852   1  99
11               15.447411   1  99
12               14.327957   1  99
13               16.548642   1  99
14               16.960241   1  99
15               16.121948   1  99
16               17.803851   1  99
17               15.593633   1  99
18               19.674252   1  99
19               16.899756   1  99
20               16.255442   1  99
21               25.535479   1  98

In [11]:
opm.head(15)

product_id                product_name  aisle_id  department_id  prices  \
0            1  Chocolate Sandwich Cookies        61             19     5.8   
1            1  Chocolate Sandwich Cookies        61             19     5.8   
2            1  Chocolate Sandwich Cookies        61             19     5.8   
3            1  Chocolate Sandwich Cookies        61             19     5.8   
4            1  Chocolate Sandwich Cookies        61             19     5.8   
5            1  Chocolate Sandwich Cookies        61             19     5.8   
6            1  Chocolate Sandwich Cookies        61             19     5.8   
7            1  Chocolate Sandwich Cookies        61             19     5.8   
8            1  Chocolate Sandwich Cookies        61             19     5.8   
9            1  Chocolate Sandwich Cookies        61             19     5.8   
10           1  Chocolate Sandwich Cookies        61             19     5.8   
11           1  Chocolate Sandwich Cookies        61             19     5.8   
12           1  Chocolate Sandwich Cookies        61             19     5.8   
13           1  Chocolate Sandwich Cookies        61             19     5.8   
14           1  Chocolate Sandwich Cookies        61             19     5.8   

    order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    3139998      138            28                   6                 11   
1    1977647      138            30                   6                 17   
2     389851      709             2                   0                 21   
3     652770      764             1                   3                 13   
4    1813452      764             3                   4                 17   
5    1701441      777            16                   1                  7   
6    1871483      825             3                   2                 14   
7    1290456      910            12                   3                 10   
8     369558     1052            10                   1                 20   
9     589712     1052            15                   1                 12   
10   1798115     1379             1                   5                 11   
11    497193     1480            36                   3                 13   
12   1727826     1480            38                   4                 13   
13   2694698     1480            39                   2                 12   
14   3171753     1494             6                   3                 10   

    days_since_prior_order  new_customer  add_to_cart_order  reordered _merge  \
0                      3.0         False                  5          0   both   
1                     20.0         False                  1          1   both   
2                      6.0         False                 20          0   both   
3                      NaN          True                 10          0   both   
4                      9.0         False                 11          1   both   
5                     26.0         False                  7          0   both   
6                     30.0         False                  2          0   both   
7                     30.0         False                  1          0   both   
8                     19.0         False                  1          0   both   
9                     15.0         False                  2          1   both   
10                     NaN          True                  1          0   both   
11                     7.0         False                  1          0   both   
12                     7.0         False                  3          1   both   
13                    12.0         False                  4          1   both   
14                     2.0         False                  8          0   both   

      price_range_loc     busiest_day    busiest_days   busiest_hours  
0   Mid-range product  Regularly busy  Regularly busy     Most orders  
1   Mid-range product  Regularly busy  Regularly 

In [12]:
opm.head(100)

product_id                product_name  aisle_id  department_id  prices  \
0            1  Chocolate Sandwich Cookies        61             19     5.8   
1            1  Chocolate Sandwich Cookies        61             19     5.8   
2            1  Chocolate Sandwich Cookies        61             19     5.8   
3            1  Chocolate Sandwich Cookies        61             19     5.8   
4            1  Chocolate Sandwich Cookies        61             19     5.8   
..         ...                         ...       ...            ...     ...   
95           1  Chocolate Sandwich Cookies        61             19     5.8   
96           1  Chocolate Sandwich Cookies        61             19     5.8   
97           1  Chocolate Sandwich Cookies        61             19     5.8   
98           1  Chocolate Sandwich Cookies        61             19     5.8   
99           1  Chocolate Sandwich Cookies        61             19     5.8   

    order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    3139998      138            28                   6                 11   
1    1977647      138            30                   6                 17   
2     389851      709             2                   0                 21   
3     652770      764             1                   3                 13   
4    1813452      764             3                   4                 17   
..       ...      ...           ...                 ...                ...   
95    602103    10831             8                   3                 11   
96     49629    11119            18                   1                 14   
97    317888    11186            13                   5                 16   
98    682486    11243            16                   3                 13   
99   2861386    11448            10                   2                 15   

    days_since_prior_order  new_customer  add_to_cart_order  reordered _merge  \
0                      3.0         False                  5          0   both   
1                     20.0         False                  1          1   both   
2                      6.0         False                 20          0   both   
3                      NaN          True                 10          0   both   
4                      9.0         False                 11          1   both   
..                     ...           ...                ...        ...    ...   
95                    23.0         False                  5          0   both   
96                    30.0         False                  1          0   both   
97                     2.0         False                  8          0   both   
98                     0.0         False                  2          0   both   
99                     3.0         False                  6          0   both   

      price_range_loc     busiest_day    busiest_days   busiest_hours  
0   Mid-range product  Regularly busy  Regularly busy     Most orders  
1   Mid-range product  Regularly busy  Regularly busy  Average orders  
2   Mid-range product     Busiest day    Busiest days  Average orders  
3   Mid-range product  Regularly busy    Slowest days     Most orders  
4   Mid-range product      Least busy    Slowest days  Average orders  
..                ...             ...             ...             ...  
95  Mid-range product  Regularly busy    Slowest days     Most orders  
96  Mid-range product  Regularly busy    Busiest days     Most orders  
97  Mid-range product  Regularly busy  Regularly busy     Most orders  
98  Mid-range product  Regularly busy    Slowest days     Most orders  
99  Mid-range product  Regularly busy  Regularly busy     Most orders  

[100 rows x 19 columns]

## 02 Aggregated mean of the “order_number” column grouped by “department_id” 

In [13]:
opm.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

## 03 Create loyalty flag for existing customers 
using the transform() and loc() functions


#### Transform function

In [14]:
# creating an new column max_order and calculating the results
opm['max_order'] = opm.groupby(['user_id'])['order_number'].transform(np.max)

#### loc fuctions

In [15]:
# Creating a red flag for customers with > 40 orders
opm.loc[opm['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [16]:
# Creating a red flag for customers with at least 11 orders and max 40
opm.loc[(opm['max_order'] <= 40) & (opm['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [17]:
# Creating a red flag for customers with 10 or less ordered products
opm.loc[opm['max_order'] <=10, 'loyalty_flag'] = 'New Customer'

In [18]:
# Checking the head for 3 columns
opm[['user_id', 'loyalty_flag', 'order_number']].head(60)

user_id      loyalty_flag  order_number
0       138  Regular customer            28
1       138  Regular customer            30
2       709      New Customer             2
3       764      New Customer             1
4       764      New Customer             3
5       777  Regular customer            16
6       825      New Customer             3
7       910  Regular customer            12
8      1052  Regular customer            10
9      1052  Regular customer            15
10     1379      New Customer             1
11     1480    Loyal customer            36
12     1480    Loyal customer            38
13     1480    Loyal customer            39
14     1494  Regular customer             6
15     1494  Regular customer             7
16     1494  Regular customer             9
17     1540  Regular customer             4
18     1540  Regular customer             5
19     1540  Regular customer             6
20     1540  Regular customer             7
21     1540  Regular customer             8
22     1540  Regular customer             9
23     1540  Regular customer            14
24     1540  Regular customer            15
25     1540  Regular customer            18
26     1540  Regular customer            19
27     1540  Regular customer            20
28     1540  Regular customer            21
29     1540  Regular customer            22
30     1540  Regular customer            24
31     1540  Regular customer            25
32     1540  Regular customer            27
33     1540  Regular customer            28
34     1598      New Customer             1
35     1598      New Customer             2
36     1598      New Customer             3
37     1647      New Customer             1
38     1647      New Customer             4
39     1647      New Customer             6
40     1647      New Customer             8
41     1647      New Customer             9
42     2833      New Customer             1
43     2850  Regular customer             7
44     2857  Regular customer            24
45     3010  Regular customer             5
46     3029      New Customer             1
47     3395      New Customer             4
48     3904      New Customer             1
49     4006    Loyal customer             2
50     4006    Loyal customer             3
51     4006    Loyal customer             5
52     4006    Loyal customer             9
53     4006    Loyal customer            12
54     4122  Regular customer            11
55     4141      New Customer             2
56     4343      New Customer             1
57     4553      New Customer             3
58     4845  Regular customer             6
59     6866      New Customer             1

In [19]:
# Counting the values of loyalty_flag
opm['loyalty_flag'].value_counts(dropna = False)

loyalty_flag
Regular customer    15876776
Loyal customer      10284093
New Customer         6243990
Name: count, dtype: int64

## 04 Descriptive statistics by loyalty
The marketing team at Instacart wants to know whether there’s a difference between the spending habits of the three types of customers you identified. Use the loyalty flag you created and check the basic statistics of the product prices for each loyalty category (Loyal Customer, Regular Customer, and New Customer). What you’re trying to determine is whether the prices of products purchased by loyal customers differ from those purchased by regular or new customers.


In [20]:
# Checking for the right column
opm.columns

Index(['product_id', 'product_name', 'aisle_id', 'department_id', 'prices',
       'order_id', 'user_id', 'order_number', 'orders_day_of_week',
       'order_hour_of_day', 'days_since_prior_order', 'new_customer',
       'add_to_cart_order', 'reordered', '_merge', 'price_range_loc',
       'busiest_day', 'busiest_days', 'busiest_hours', 'max_order',
       'loyalty_flag'],
      dtype='object')

In [21]:
# Grouping the data within the loyalty_flags and calculate the mean
opm.groupby('loyalty_flag').agg({'prices': ['mean', 'max']})

prices         
                       mean      max
loyalty_flag                        
Loyal customer    10.386336  99999.0
New Customer      13.294670  99999.0
Regular customer  12.495717  99999.0

#### In average loyal customer buy less expensive products than New and Regular customers represented by an mean of 10.39$ New Customers spend in average 13.29$ on their products, followed by regular customers with an avergae of 12.50$

## 05 Create spending_flag
The team now wants to target different types of spenders in their marketing campaigns. This can be achieved by looking at the prices of the items people are buying. Create a spending flag for each user based on the average price across all their orders using the following criteria:

    If the mean of the prices of products purchased by a user is lower than 10, then flag them as a “Low spender.”
    If the mean of the prices of products purchased by a user is higher than or equal to 10, then flag them as a “High spender.”


In [22]:
# Creating a new column mean_price
opm['mean_price']= opm.groupby(['user_id'])['prices'].transform('mean')

In [23]:
opm.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   3139998      138            28                   6                 11   
1   1977647      138            30                   6                 17   
2    389851      709             2                   0                 21   
3    652770      764             1                   3                 13   
4   1813452      764             3                   4                 17   

   ...  add_to_cart_order  reordered  _merge    price_range_loc  \
0  ...                  5          0    both  Mid-range product   
1  ...                  1          1    both  Mid-range product   
2  ...                 20          0    both  Mid-range product   
3  ...                 10          0    both  Mid-range product   
4  ...                 11          1    both  Mid-range product   

      busiest_day    busiest_days   busiest_hours max_order      loyalty_flag  \
0  Regularly busy  Regularly busy     Most orders        32  Regular customer   
1  Regularly busy  Regularly busy  Average orders        32  Regular customer   
2     Busiest day    Busiest days  Average orders         5      New Customer   
3  Regularly busy    Slowest days     Most orders         3      New Customer   
4      Least busy    Slowest days  Average orders         3      New Customer   

   mean_price  
0    6.935811  
1    6.935811  
2    7.930208  
3    4.972414  
4    4.972414  

[5 rows x 22 columns]

In [24]:
# Creating a new column 'spendings'
opm.loc[opm['mean_price'] < 10, 'spendings'] = 'Low spender'

In [25]:
# Creating a new column 'spendings'
opm.loc[opm['mean_price'] >= 10, 'spendings'] = 'High spender'

In [26]:
# Checking 'spendings' value_counts
opm['spendings'].value_counts()

spendings
Low spender     31770614
High spender      634245
Name: count, dtype: int64

## 06 Create order_frequency flag
In order to send relevant notifications to users within the app (for instance, asking users if they want to buy the same item again), the Instacart team wants you to determine frequent versus non-frequent customers. Create an order frequency flag that marks the regularity of a user’s ordering behavior according to the median in the “days_since_prior_order” column. The criteria for the flag should be as follows:

    If the median of “days_since_prior_order” is higher than 20, then the customer should be labeled a “Non-frequent customer.”
    If the median is higher than 10 and lower than or equal to 20, then the customer should be labeled a “Regular customer.”
    If the median is lower than or equal to 10, then the customer should be labeled a “Frequent customer.”


In [27]:
# Creating a column 'frequency median'
opm['frequency_median']=opm.groupby(['user_id'])['days_since_prior_order'].transform('median')

In [28]:
# Checking dataframes head
opm.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   3139998      138            28                   6                 11   
1   1977647      138            30                   6                 17   
2    389851      709             2                   0                 21   
3    652770      764             1                   3                 13   
4   1813452      764             3                   4                 17   

   ...  _merge    price_range_loc     busiest_day    busiest_days  \
0  ...    both  Mid-range product  Regularly busy  Regularly busy   
1  ...    both  Mid-range product  Regularly busy  Regularly busy   
2  ...    both  Mid-range product     Busiest day    Busiest days   
3  ...    both  Mid-range product  Regularly busy    Slowest days   
4  ...    both  Mid-range product      Least busy    Slowest days   

    busiest_hours max_order      loyalty_flag mean_price    spendings  \
0     Most orders        32  Regular customer   6.935811  Low spender   
1  Average orders        32  Regular customer   6.935811  Low spender   
2  Average orders         5      New Customer   7.930208  Low spender   
3     Most orders         3      New Customer   4.972414  Low spender   
4  Average orders         3      New Customer   4.972414  Low spender   

   frequency_median  
0               8.0  
1               8.0  
2               8.0  
3               9.0  
4               9.0  

[5 rows x 24 columns]

In [29]:
# Creating a new column order_frequency and assigning 'Non-frequent customer'

opm.loc[opm['frequency_median'] > 20, 'order_frequency'] = 'Non-frequent customer'

In [30]:
# Creating a new column order_frequency and assigning 'Regular customer'
opm.loc[(opm['frequency_median'] <= 20) & (opm['frequency_median'] > 10), 'order_frequency'] = 'Regular customer'

In [31]:
# Creating a new column order_frequency and assigning 'Frequent customer'

opm.loc[opm['frequency_median'] <= 10, 'order_frequency'] = 'Frequent customer'

In [32]:
# Checking 'order_frequency's value_counts
opm['order_frequency'].value_counts()

order_frequency
Frequent customer        21559853
Regular customer          7208564
Non-frequent customer     3636437
Name: count, dtype: int64

## 07 Export data

In [34]:
# Export data
opm.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_with_flags.pkl'))